### PR TITLE
[skip ci]Feature/pipeline masterci

### PR DESCRIPTION
**Pull request by PengTian0**
Put Jenkins Job "[MasterCI](http://rackhdci.lss.emc.com/job/BuildRelease/job/MasterCI/)" under source code control.

Test:
Now, the test job has 3 stable running and the failed one is caused by failure  cases of continuous test.
image

A successful running of the pipeline:
image

### CODE REVIEW

**Comment by PengTian0** [comment]
@panpan0000 @changev @keedya

File: jobs/build_ova/build_ova.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+set -x
+
+
+cd $WORKSPACE/build/packer 
+
+export PACKER_CACHE_DIR=/tmp/packer_cache
+export BUILD_TYPE=vmware
+
+#exprot vars to build ova
```
Comment (anhou) [inline jobs/build_ova/build_ova.sh]: typo export

File: jobs/build_ova/build_ova.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    export ANSIBLE_PLAYBOOK=rackhd_release
+else
+    export ANSIBLE_PLAYBOOK=rackhd_ci_builds
+fi
+export RACKHD_VERSION=$RACKHD_VERSION
+#export end
+
+./HWIMO-BUILD
+
+#cd $WORKSPACE/build/packer
```
Comment (anhou) [inline jobs/build_ova/build_ova.sh]: all comment out sentences could be removed except there're some other purposes which also need to be noted

File: jobs/build_ova/ova_post_test.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash +xe
+
+#fix bug
```
Comment (anhou) [inline jobs/build_ova/ova_post_test.sh]: fix what bug?

File: jobs/build_ova/ova_post_test.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+delete_ova() {
+	ansible esxi -a "./vm_operation.sh -a delete ${ESXI_HOST_IP} 1 ova-for-post-test"
```
Comment (anhou) [inline jobs/build_ova/ova_post_test.sh]: indent

File: jobs/build_vagrant/build_vagrant.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+set +e
+
+packer -v
+vagrant -v
+
+cd $WORKSPACE/build/packer/ansible/roles/rackhd-builds/tasks
+
+
+############# DEBUG ######################20160117  for 1.0.0 release
```
Comment (anhou) [inline jobs/build_vagrant/build_vagrant.sh]: could be removed

File: jobs/build_vagrant/build_vagrant.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+cd ..
+
+cd $WORKSPACE/build/packer 
+
+
+export PACKER_CACHE_DIR=/tmp/packer_cache
+
+
+#exprot vars to build virtualbox
```
Comment (anhou) [inline jobs/build_vagrant/build_vagrant.sh]: typo

File: jobs/release/release_docker.sh [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+#move the original clean up steps to post-build, avoid failure impacts build status.
+set +e
+set -x
+
+
+
+clean_up(){
+# parameter : the images keyword to be delete
+keyword=$1
```
Comment (anhou) [inline jobs/release/release_docker.sh]: no indent in this function

File: jobs/build_ova/build_ova.sh (hunk 1 above)
Comment (PengTian0) [reply jobs/build_ova/build_ova.sh]: Thanks!
Fixed!

File: jobs/build_ova/build_ova.sh (hunk 2 above)
Comment (PengTian0) [reply jobs/build_ova/build_ova.sh]: Thanks!
Fixed!

File: jobs/build_ova/ova_post_test.sh (hunk 3 above)
Comment (PengTian0) [reply jobs/build_ova/ova_post_test.sh]: It's just an old comment.
And I removed it.

File: jobs/build_ova/ova_post_test.sh (hunk 4 above)
Comment (PengTian0) [reply jobs/build_ova/ova_post_test.sh]: Thanks!
Fixed!

File: jobs/build_vagrant/build_vagrant.sh (hunk 5 above)
Comment (PengTian0) [reply jobs/build_vagrant/build_vagrant.sh]: Thanks!
Fixed!

File: jobs/build_vagrant/build_vagrant.sh (hunk 6 above)
Comment (PengTian0) [reply jobs/build_vagrant/build_vagrant.sh]: Thanks!
Fixed!

File: jobs/release/release_docker.sh (hunk 7 above)
Comment (PengTian0) [reply jobs/release/release_docker.sh]: Thanks!
Fixed!

File: jobs/MasterCI/MasterCI [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+           load("jobs/build_ova/ova_post_test.groovy")
+
+        }, 'docker post test':{
+
+           load("jobs/build_docker/docker_post_test.groovy")
+
+        }
+    }
+
+    stage("Release"){
```
Comment (panpan0000) [inline jobs/MasterCI/MasterCI]: I prefer this stage name as "publish" or "upload". the "release" term is too generic :-)

File: jobs/build_docker/build_docker.groovy [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                deleteDir()
+                def current_workspace = pwd()
+                def shareMethod
+                dir("Build_Docker_JFiles"){
+                    checkout scm
+                    shareMethod = load("jobs/shareMethod.groovy")
+                }
+ 
+                def url = "https://github.com/changev/on-build-config.git"
+                def branch = "*/workaround/commit_hash_as_pack_refs_and_1_run_of_apt_get_update"
```
Comment (panpan0000) [inline jobs/build_docker/build_docker.groovy]: will this workaround being removed ?
once @changev;s PR merged

File: jobs/build_debian/build_debian.groovy [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                    checkout scm
+                    shareMethod = load("jobs/shareMethod.groovy")
+                }
+                def url = "https://github.com/RackHD/on-build-config.git"
+                def branch = "*/master"
+                def targetDir = "on-build-config"
+                shareMethod.checkout(url, branch, targetDir)
+
+                withCredentials([
+                    usernameColonPassword(credentialsId: "ff7ab8d2-e678-41ef-a46b-dd0e780030e1",
```
Comment (panpan0000) [inline jobs/build_debian/build_debian.groovy]: Nice. 
the credentialsId is binding to Jenkins server. (aka, it's not quite portable if community run this groovy in their Jenkins. so at least add comments in code. )

File: jobs/build_docker/build_docker.sh [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+cp $output_path/builds/* common/
+cp $output_path/syslinux/* pxe/
+cp $output_path/ipxe/* pxe/
+sudo chown -R $USER:$USER common
+sudo chown -R $USER:$USER pxe
+
+popd
+
+#docker images build
+#test(){
```
Comment (panpan0000) [inline jobs/build_docker/build_docker.sh]: please ask @changev to remove the debug code

File: jobs/build_ova/build_ova.sh [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+set +e
+#sudo cp ${HOME}/bin/packer   /usr/bin
+#sudo apt-get install -y  jq
+
+cd $WORKSPACE/build/packer/ansible/roles/rackhd-builds/tasks
+sed -i "s#https://dl.bintray.com/rackhd/debian trusty release#https://dl.bintray.com/rackhd-mirror/debian trusty main#" main.yml
+sed -i "s#https://dl.bintray.com/rackhd/debian trusty main#https://dl.bintray.com/rackhd-mirror/debian trusty main#" main.yml
+cd ..
+
+pkill packer
```
Comment (panpan0000) [inline jobs/build_ova/build_ova.sh]: I still have it in script ?!
aha, should I use signal handler ? :-P

File: jobs/build_ova/ova_post_test.sh [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+}
+delete_ova
+sleep 10
+
+PACKERDIR="$WORKSPACE/build/packer/"
+OVA="$PACKERDIR/rackhd-${OS_VER}-${RACKHD_VERSION}.ova"
+
+bash ./build-config/build-release-tools/post_test.sh \
+--type ova \
+--adminIP 10.62.59.167 \
```
Comment (panpan0000) [inline jobs/build_ova/ova_post_test.sh]: [!] Don't expose IP !!!!!

File: jobs/release/release_vagrant.sh [hunk 14]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+
+set +e
+echo "upload vagrant to atlas"
+# for test
+cd VAGRANT
+mv $(ls) $(ls | sed "s/1.1-1/1.1.1/")
```
Comment (panpan0000) [inline jobs/release/release_vagrant.sh]: is it a temp workaround?

**Comment by panpan0000** [comment]
👍  my other suggestions are nice-to-have comments.

File: jobs/release/release_vagrant.sh (hunk 14 above)
Comment (changev) [reply jobs/release/release_vagrant.sh]: can be deleted now

File: jobs/build_docker/build_docker.sh [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  checkout \
+  packagerefs-commit
+
+rsync -r ./$CLONE_DIR/RackHD/ ./build/
+
+
+
+#build static files
+pushd ./$CLONE_DIR/on-imagebuilder
+#sudo ./build_all.sh
```
Comment (changev) [inline jobs/build_docker/build_docker.sh]: should remove #

File: jobs/build_docker/build_docker.sh (hunk 11 above)
Comment (changev) [reply jobs/build_docker/build_docker.sh]: This was added by @PengTian0 for debug workflow

File: jobs/MasterCI/MasterCI [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        }, 'build debian':{
+                load("jobs/build_debian/build_debian.groovy")
+
+        }, 'build docker':{
+            
+               load("jobs/build_docker/build_docker.groovy")
+        }
+    }
+
+    stage("Vagrant & OVA Build"){
```
Comment (anhou) [inline jobs/MasterCI/MasterCI]: Could it be 'VM Images Build' ? which more generic than specific VM type

Comment (PengTian0) [reply jobs/MasterCI/MasterCI]: Thanks!
Fixed.

File: jobs/build_docker/build_docker.sh (hunk 15 above)
Comment (PengTian0) [reply jobs/build_docker/build_docker.sh]: Thanks!
Fixed.

File: jobs/release/release_vagrant.sh (hunk 14 above)
Comment (PengTian0) [reply jobs/release/release_vagrant.sh]: Thanks!
Fixed.

File: jobs/build_docker/build_docker.sh (hunk 11 above)
Comment (PengTian0) [reply jobs/build_docker/build_docker.sh]: Thanks!
Fixed.

File: jobs/build_docker/build_docker.groovy (hunk 9 above)
Comment (PengTian0) [reply jobs/build_docker/build_docker.groovy]: Thanks!
Fixed.

File: jobs/MasterCI/MasterCI (hunk 8 above)
Comment (PengTian0) [reply jobs/MasterCI/MasterCI]: Thanks!
Fixed.

**Comment by amymullins** [comment]
-1 This is a great step in the right direction but comments below: 

Building release packages should only take place after testing has passed.  You don't want to consume system resources on building up packages that are known to fail test.  If time is a concern, we should review the steps taken to build the debian and docker packages.

Request to move the Debian and docker builds to a generic "package build" stage that includes the OVA and Vagrant images.

For error handling, what happens between each stage when a failure occurs?  Will the pipeline fail?  Examples:
If the "Create Manifest" stage fails, is does the pipeline go red and abort the test?
Same question for the "build and test" stage, if that stage fails will the pipeline go red and abort?

**Comment by panpan0000** [comment]
**Question**
>[@amymullins] Request to move the Debian and docker builds to a generic "package build" stage that includes the OVA and Vagrant images.
**Answer:** : 
it's because the OVA/Vagrant are installed from debian packages, the deb packages exactly reflects the content of the manifest. although there's ```rackhd_local.yml``` ansible in rackhd/packer , but it can't accept a manifest for now... by the way, we can also covered the validation of  both install-from-deb and install-from-src (  docker build is install-from-source-code) (aka, if deb packages are bad, ova-build may fail.)
If the timing is  concern and this indeed needs being address,
that will be not small efforts to make it parallel in my mind:
1. build up another way to verify the debian packages.( I can't figure out a better way than installing them..)
2. refactor packer scripts(install-from-src) to make it accept manifest.

my option is : this PR is about transition from "traditional Jenkins job" to "Jenkins pipeline".without this PR, those Jenkins job still runs in the same logic and flow.
so as agile point of view, **those enhancement can happen later** , because of effort.

File: jobs/build_debian/build_debian.sh [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+--jobs 8 \
+--force \
+--is-official-release $IS_OFFICIAL_RELEASE
+
+
+
+
+./on-build-config/build-release-tools/HWIMO-BUILD on-build-config/build-release-tools/application/release_debian_packages.py \
+--build-directory b/ \
+--bintray-credential BINTRAY_CREDS \
```
Comment (changev) [inline jobs/build_debian/build_debian.sh]: missing $?

**Comment by PengTian0** [comment]
@amymullins 
Your ideas are great, most make good sense  with BRI careful evaluation and analytic .
And I agree we can take this opportunity to make those optimizations in near term. Here’s the epic [RAC-4259](https://rackhd.atlassian.net/browse/RAC-4259) to track your great ideas.

But Rome was not built in a day, as agile perspective, one PR can’t achieve everything in one time.
Taken this PR focus on “pipeline” Jenkins jobs. It just turns original Jenkins logic into pipeline code and make it source controlled . without this PR, those Jenkins job still runs in the same logic and flow.
_Just assuming there’s a story about re-writing a pcs of code from C++ to Java , it doesn’t require to fix original C++ bugs in that story._

Taken there’s additional effort, in Scrum spirit,  do you agree we have an incremental implementation plan ? 

**1. Build after Test pass.  Agree.**  
But we can’t directly move build steps after test steps. Either OVA, vagrant, docker requires on-imagebuilder static files( it was created in debian-build step ). Now it’s build parallel with tests steps, if simply doing the moving,  the total time will become longer (like below chart (Before v.s. After))
So [RAC-4261](https://rackhd.atlassian.net/browse/RAC-4261) is tracked for move build after test pass. Which depends on on-imagebuilder-build optimization [RAC-4260](https://rackhd.atlassian.net/browse/RAC-4260)

image


**2. Request to move the Debian and docker builds to a generic "package build" stage that includes the OVA and Vagrant images.**
It’s by design.  We will have to think deeper. 
Current OVA/Vagrant build depend on previous debian-build step. Because inside OVA/Vagrant, RackHD is installed from deb packages.
Rational:
2.1   Otherwise, we will have to install from source for ova/vagrant, you know, there’s time window problem( new commit during smoke-test steps)

2.2   “install from deb” as a typical user-scenario, doing it during ova/vagrant build also covers the validation:  whether those deb we build can be successfully installed and run. (of course, “docker build” verifies  “install-from-src”)

2.3   Side-effect Impact:   currently, we manage RackHD via “service” (like ```service rackhd start```) , But install-from-src lost this capability.

2.4   Same as above, on-imagebuilder takes long time, so doing debian-build at the same time with ova-build/vagrant-build won’t save total time. depends on RAC-4260

2.5   Moreover, the ROI:  will the “R” so significant comparing the “I” , that we need to do it immediately ?
to make it simple, without considering on-imagebuilder,    as below chart. Make them parallel will save time for two “orange” blocks( pack deb & install from deb.) (refer to last chart)
If time of two orange blocks is small , the Return in ROI will be re-considered.
We haven’t got time to verify how long orange blocks take (current install from deb is not fast because slave in SH, and network is slow. Need to test on US machine ). That’s a TBD

2.6  Efforts needed (The Investment of ROI) , when assuming we make deb-build & ova-build parallel.
a) Packer Change: refactor packer scripts(install-from-src) to make it accept a manifest. To address the “time window” problem.
b) Add debian-post-build step:  build up another way to verify the debian packages.( like vagrant-post-test,  vagrant up then install deb, then check service/api )

image

                
**3. Error Handling**

The pipeline will fail and abort if one stage fails.

File: jobs/build_debian/build_debian.sh (hunk 17 above)
Comment (PengTian0) [reply jobs/build_debian/build_debian.sh]: No~ 
The script release_debian_packages.py will read the value of the environment variable.

**Comment by panpan0000** [comment]
agree. @PengTian0 .  let's use those stories to address @amymullins's suggestion, which falls into optimization of the CICD logic .  this PR focus on turn existing Jenkins logic into pipeline groovy.

**Comment by amymullins** [comment]
Discussed earlier today, this was an agreeable baseline to introduce a test pipeline and Jenkins jobs under source control.  Additional optimizations and enhancements can iterate from this baseline.  This is a multi-team effort to put these efficiencies in place in a short timeframe.

Request @RackHD/corecommitters @keedya  and QRB team members  ( @tannoa2 , @pscharla , @tldavies , @hohene , @changev ) to help with a technical review.

File: jobs/create_manifest/create_manifest.groovy [hunk 18]
```diff
@@ -0,0 +1,43 @@
+node(create_manifest_node){
```
Comment (changev) [inline jobs/create_manifest/create_manifest.groovy]: -1 for this stage.
this stage has deleteDir(), but doesn't use a new workspace.
Cause pipeline job masterci is not be restricted on specific nodes, so if this stage runs on the same node with masterci pipeline job, then the workspace of masterci will be deleted.
subsequent stages can't find their groovy scipts and will fail.

**Comment by changev** [comment]
:+1:

**Comment by keedya** [comment]
Pipeline scripts look good. Only one comment; I don't think we need all most of the shell scripts. I believe these script can be converted as well to groovy scripts
👍

File: jobs/build_debian/build_debian.groovy [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+                dir("Build_Debian_JFiles"){
+                    checkout scm
+                    shareMethod = load("jobs/shareMethod.groovy")
+                }
+                def url = "https://github.com/RackHD/on-build-config.git"
+                def branch = "*/master"
+                def targetDir = "on-build-config"
+                shareMethod.checkout(url, branch, targetDir)
+
+                // credentials  are bingding to Jenkins Server
```
Comment (anhou) [inline jobs/build_debian/build_debian.groovy]: typo

File: jobs/MasterCI/MasterCI [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    
+    stage("Create Manifest"){
+        load("jobs/create_manifest/create_manifest.groovy")
+        def manifest_name=env.MANIFEST_FILE_URL.tokenize('/')[-1];
+        currentBuild.description = "<a href=${env.MANIFEST_FILE_URL}>${manifest_name}</a>";
+    }
+
+    stage("Build & Test"){
+
+	parallel 'Continus Functional Test':{
```
Comment (anhou) [inline jobs/MasterCI/MasterCI]: indent, and many unintended blank lines. please clean them

**Comment by anhou** [comment]
👍  from me

File: jobs/MasterCI/MasterCI (hunk 20 above)
Comment (PengTian0) [reply jobs/MasterCI/MasterCI]: Thanks!
Fixed

File: jobs/build_debian/build_debian.groovy (hunk 19 above)
Comment (PengTian0) [reply jobs/build_debian/build_debian.groovy]: Thanks!
Fixed

**Comment by PengTian0** [comment]
@keedya 
Thank you so much for your reviewing!
Yes, I can convert the shell script to groovy, but there will be so much indent in the groovy.
 So I just put the long shell command in shell script!

**Comment by keedya** [comment]
👍

**Pull request merged**